Combined schema index provider

tinwelint opened the pull request:
Basically a `SchemaIndexProvider` that combines two SchemaIndexProviders, one native provider with a limited set of capabilities (numbers for now) and one to fall back everything else to (the standard lucene provider for now).

The combination starts at the `SchemaIndexProvider` and extends all the way out to the `IndexPopulator`, `IndexAccessor` and its associates. For this to work a `NativeSchemaNumberIndexProvider` is also implemented here.

This PR does not yet introduce this combined provider as a loadable kernel extension so no migration or change in default provider is included right now.

tinwelint committed:
Avoids flushing index if no changes takes place in updater

tinwelint committed:
Embryo of combined SchemaIndexProvider

just setting the general structure, or trying it out rather.

Commit:
Tests for CombinedSchemaIndexProvider

Commit:
Tests for CombinedIndexAccessor

Commit:
Tests for CombinedIndexPopulator

Commit:
Introduce test helper for combined index tests

Commit:
Test for CombinedIndexReader

Commit:
Better test for close in CombinedIndex

Commit:
Cleanup combined index tests

Commit:
Tests of CombinedIndexUpdater

Commit:
Checkstyle

Commit:
Implement query for CombinedIndexReader

Commit:
Simplify combined schema index provider 'select'

Commit:
Implement getAllEntriesReader for CombinedIndexAccessor

Commit:
NativeSchemaIndexProvider: Step1

- getPopulator
- getOnlineAccessor
- getPopulationFailure

Note: new 'readHeader' functionality in GBPTree to be
able to read stored header without opening the tree
for writes.

tinwelint committed:
Javadoc and de-generified ReadOnlyMetaLayout

tinwelint committed:
Implements NativeSchemaIndexProvider.getInitialState

tinwelint committed:
Implemented migration participants as NOT_PARTICIPATING for the time being

tinwelint committed:
GBPTreeTest refactorings after rebase

tinwelint committed:
Don't test with null string array

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/Layout.java:
```diff
@@ -161,16 +163,34 @@ static long namedIdentifier( String name, int checksum )
+     *     <li>{@code minorVersion == this.minorVersion()}</li>
+     * </ul>
+     * <p>
+     * Edge case: When only want to {@link GBPTree#readHeader(PageCache, File, Layout, Header.Reader)} read header} of index and not
```
not really clear what this edge case describing, why they are equal, what is the contract of those qualities, should all layouts be compatible with those requirements?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/combined/CombinedSchemaIndexProvider.java:
```diff
@@ -0,0 +1,135 @@
+            return fallback;
+        }
+
+        Value singleValue = values[0];
```
in case if this single value is also part of composite indexes definition and now only one property is updated we should take that into account when we select proper index implementation

tinwelint replied:
Index updates always arrive to the index "complete", so even if there's a node N with label L and property keys A and B. Let's also say that the index for L and A and B exists and N has values in A and B... updating just A or B will still produce an update to the index which consists of both A and B. So this is a case we don't need to worry about.

burqen replied:
We just talked about this but just want to type it out for completeness of this PR.

All index updates come as complete `IndexEntryUpdate` objects which contain the whole index entry that has been updated, not only the single part that is actually changed. So this should not be a problem.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/combined/CombinedIndexPopulator.java:
```diff
@@ -0,0 +1,142 @@
+    @Override
+    public void includeSample( IndexEntryUpdate update )
+    {
+        boostPopulator.includeSample( update );
```
Should we also propagate update only to correct populator right? otherwise, it will be doubled in the end?

tinwelint replied:
Yes, I think you're right about that.

MishaDemianenko replied:
this one is still there and include update in both samples

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/combined/CombinedSchemaIndexProvider.java:
```diff
@@ -0,0 +1,135 @@
+        return fallback;
+    }
+
+    static IndexSample combineSamples( IndexSample first, IndexSample other )
```
would move that to CombineIndexSampler instead

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/combined/CombinedSchemaIndexProvider.java:
```diff
@@ -0,0 +1,135 @@
+import static java.lang.String.format;
+
+/**
+ * Exactly what this provider will be called and where it will live is undecided, so for the time being
```
this part of javadoc is not that useful and should go

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/combined/CombinedSchemaIndexProvider.java:
```diff
@@ -0,0 +1,135 @@
+    private final SchemaIndexProvider boostProvider;
+    private final SchemaIndexProvider fallbackProvider;
+
+    public CombinedSchemaIndexProvider( SchemaIndexProvider boostProvider, SchemaIndexProvider fallbackProvider )
```
I think already here as soon as we know that index is combined based on provided descriptor we should just return lucene index implementations

burqen replied:
There are a few alternatives:

1. Lucene and Combined providers living side by side.
What provider is used when an index is created will then look something like:
```
if ( compositeIndex ) { return LuceneProvider }
if ( configuredToOnlyUseLucene ) { return LuceneProvider }
return CombinedProvider
```
In this way, the Combined index will only be used when there is actually a chance that it will see updates for Native part.

2. We only have Combined provider (like what this PR assumes)
Then it will need to handle the case for composite keys as well (by sending them all to Lucene).

What route we will take is still undecided and the this code will adapt when a decision is taken.

Did this make it more clear?

tinwelint replied:
There's still benefit in creating an index as "composite lucene native" instead of lucene, and that is that the native provider will over time be able to handle more and more value types so having all indexes as composite will see "automatic" benefits over time, but lucene will not.

tinwelint replied:
If we have a configuration option and it's set to only use lucene then I don't think that the composite provider should be even involved, do you?

tinwelint replied:
If we have a configuration option and it's set to only use lucene then I don't think that the composite provider should be even involved, do you?

MishaDemianenko replied:
i think it still makes sense to have a FusionSchemaIndexProvider but as soon as we know  based on index descriptor that index is combined we should just return pure lucene implementations of accessor's, populators etc since there is not a lot of value in wrapping those things right?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexHeaderReader.java:
```diff
@@ -0,0 +1,46 @@
+            short messageLength = headerData.getShort();
+            byte[] failureMessageBytes = new byte[messageLength];
+            headerData.get( failureMessageBytes );
+            failureMessage = new String( failureMessageBytes, Charset.forName( "UTF-8" ) );
```
StandardCharsets.UTF_8

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NativeSchemaNumberIndexProvider.java:
```diff
@@ -0,0 +1,181 @@
+        try
+        {
+            String failureMessage = readPopulationFailure( indexId );
+            if ( failureMessage == null )
```
do we really need to do that? maybe empty string instead?

burqen replied:
We chose this way to align with how `LuceneSchemaIndexProvider` does it.

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/combined/CombinedIndexPopulatorTest.java:
```diff
@@ -0,0 +1,379 @@
+        } );
+    }
+
+    // includeSample
```
looks like some parts are still missing

MishaDemianenko replied:
those things are still not there

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/storable/ArrayValue.java:
```diff
@@ -24,7 +24,7 @@
  * Array of one of the storable primitives
  */
-abstract class ArrayValue extends Value implements SequenceValue
+public abstract class ArrayValue extends Value implements SequenceValue
```
not sure why this became public?

MishaDemianenko reviewed (changes requested):
Looks good in general, have only a couple of comments mainly because of composite indexes.
Not sure i really like boost/fallback naming convention i would prefer if those things would be called native and lucene instead to highlight what is actually expected to be there

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NativeSchemaNumberIndexProvider.java:
```diff
@@ -115,14 +121,35 @@ public String getPopulationFailure( long indexId ) throws IllegalStateException
+        }
+        catch ( IOException e )
+        {
+            return InternalIndexState.POPULATING;
```
`LuceneSchemaIndexProvider` also do `log.error( "Failed to open index:" + indexId + ", requesting re-population.", e );`. Maybe native should as well?

Also, missing test for this behaviour.

tinwelint replied:
Yup, may be a good idea

tinwelint committed:
Renamed combined index provider to fusion

while keeping it generic, even more generic than before by passing in
a selector between one and the other. The idea is that specific combinations
of providers will live in specific kernel extensions. To have the fusion
provider generic is nice from a single responsibility perspective where
it has no idea about what the providers are, it just combines the two.

Commit:
Various small cleanups

tinwelint committed:
Checkstyle

tinwelint committed:
Fixed mocking problem in FusionSchemaIndexProviderTest

tinwelint committed:
Logs about failure to open index

tinwelint commented:
@MishaDemianenko all comments should be addressed now

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexAccessor.java:
```diff
@@ -0,0 +1,149 @@
+    public void verifyDeferredConstraints( PropertyAccessor propertyAccessor )
+            throws IndexEntryConflictException, IOException
+    {
+        nativeAccessor.verifyDeferredConstraints( propertyAccessor );
```
just a note for question, not sure it actually will work as it is now for native index

tinwelint replied:
It works fine, because for native index is a no-op

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexPopulator.java:
```diff
@@ -0,0 +1,144 @@
+    @Override
+    public void includeSample( IndexEntryUpdate update )
+    {
+        nativePopulator.includeSample( update );
```
sample still included into both populators?

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexPopulatorTest.java:
```diff
@@ -0,0 +1,381 @@
+        } );
+    }
+
+    // includeSample
```
missing tests still

tinwelint committed:
Corrects includeSample in FusionIndexPopulator

tinwelint commented:
@MishaDemianenko updated
